### PR TITLE
Expose expression for observing items in dict/list/set

### PR DESCRIPTION
**Pull request by kitchoi**
This PR exposes the `list_items`, `dict_items` and `set_items` for the observer's expression and implements the `"items"` keyword in the mini-language. All of these will allow the users to specify items in list/dict/set to be observed.

The "items" in the mini-language is translated as the union of (1) trait named "items", (2) items in list, (3) items in set, (4) items in dict.  Due to the ambiguity, these elements have `optional` flag set to true, and therefore they are less strict than using `list_items(optional=False)` (and friends) from `Expression` directly.

**Checklist**
- [x] Tests
- [x] Update API reference (`docs/source/traits_api_reference`): autodoc handles it.
- ~Update User manual (`docs/source/traits_user_manual`)~: Will be in a separate PR
- ~Update type annotation hints in `traits-stubs`~

### CODE REVIEW

File: traits/observers/expression.py [hunk 1]
```diff
@@ -10,9 +10,18 @@
...
+)
 from traits.observers._named_trait_observer import (
     NamedTraitObserver as _NamedTraitObserver,
 )
+from traits.observers._set_item_observer import (
+    SetItemObserver as _SetItemObserver,
+)
 from traits.observers._observer_graph import (
     ObserverGraph as _ObserverGraph,
 )
```
Comment (kitchoi) [inline traits/observers/expression.py:27]: Now I am less a fan of aliasing because we are going to have an `api.py` and this is a bit of nuisance. The top-level functions in this module will be exposed in the `api.py`. I am collecting this sort inconsequential renaming at the end.

File: traits/observers/expression.py [hunk 2]
```diff
@@ -196,11 +276,88 @@ def join_(*expressions):
...
+    return SingleObserverExpression(observer)
+
+
 def trait(name, notify=True, optional=False):
     """ Create a new expression for observing a trait with the exact
     name given.
 
-    Events emitted (if any) will be instances of ``TraitChangeEvent``.
+    Events emitted (if any) will be instances of
+    :py:class:`~traits.observers.events.TraitChangeEvent`.
```
Comment (kitchoi) [inline traits/observers/expression.py:360]: Drive-by improvement so that sphinx creates a link for this.

File: traits/observers/expression.py (hunk 1 above)
Comment (kitchoi) [reply traits/observers/expression.py:27]: Now that `api.py` in master, this PR should expose the new expression top-level function there too.

Comment (mdickinson) [reply traits/observers/expression.py:27]: Yes, my preference would be not to do this aliasing. We don't bother with it elsewhere in Traits, and I don't think we've ever found a reason to.

File: traits/observers/expression.py [hunk 3]
```diff
@@ -76,11 +85,82 @@ def then(self, expression):
...
+        **values** of the dict will be given to the next item in the
+        expression. For example, the following observes a trait named "number"
+        on any object that is one of the values in the dict named "mapping"::
+
+            trait("mapping").dict_items().trait("number")
+
+        Parameters
+        ----------
+        notify : boolean, optional
+            Whether to notify for changes.
```
Comment (mdickinson) [inline traits/observers/expression.py:104]: Please mention the default in the docstring (also applies to `optional` and to similar methods below).

File: traits/observers/expression.py [hunk 4]
```diff
@@ -76,11 +85,82 @@ def then(self, expression):
         """
         return SeriesExpression(self, expression)
 
+    def dict_items(self, notify=True, optional=False):
+        """ Create a new expression for observing items inside a dict.
+
+        Events emitted (if any) will be instances of
+        :py:class:`~traits.observers.events.DictChangeEvent`.
```
Comment (mdickinson) [inline traits/observers/expression.py:92]: The default Sphinx domain is `py`, so we can leave out the `:py` prefix if that makes the docstring more readable.

File: traits/observers/expression.py [hunk 5]
```diff
@@ -76,11 +85,82 @@ def then(self, expression):
...
+        to a list named ``containers``.
+
+        e.g. ``trait("containers").list_items().trait("value")`` for observing
+        the trait ``value`` on any items in the list ``containers``.
+
+        Parameters
+        ----------
+        notify : boolean, optional
+            Whether to notify for changes.
+        optional : booleal, optional
```
Comment (mdickinson) [inline traits/observers/expression.py:130]: typo: `boolean`. (Incidentally, what's the rationale for using `boolean` rather `bool` here and elsewhere? `bool` is more standard and common. I know we're duck-typing, so technically we accept anything bool-like, but I'm not sure there's value in making that distinction at this level.)

File: traits/observers/expression.py [hunk 6]
```diff
@@ -76,11 +85,82 @@ def then(self, expression):
...
+        """
+        return self.then(dict_items(notify=notify, optional=optional))
+
+    def list_items(self, notify=True, optional=False):
+        """ Create a new expression for observing items inside a list.
+
+        Events emitted (if any) will be instances of
+        :py:class:`~traits.observers.events.ListChangeEvent`.
+
+        e.g. ``trait("containers").list_items()`` for observing to mutations
```
Comment (mdickinson) [inline traits/observers/expression.py:120]:
```suggestion
        e.g. ``trait("containers").list_items()`` for observing mutations
```

File: traits/observers/expression.py [hunk 7]
```diff
@@ -196,11 +276,88 @@ def join_(*expressions):
...
+    Parameters
+    ----------
+    notify : boolean, optional
+        Whether to notify for changes.
+    optional : booleal, optional
+        Whether to ignore this if the upstream object is not a set.
+
+    Returns
+    -------
+    new_expression : traits.observers.expressions.Expression
```
Comment (kitchoi) [inline traits/observers/expression.py:349]: Oops, this `"expressions"` needs to be changed to `"expression"`.

Comment (mdickinson) [reply traits/observers/expression.py:349]: I see a few cases of `expressions` in this module, and one in `observe.py`.

Comment (kitchoi) [reply traits/observers/expression.py:349]: Thanks!  It is on my list for a final sweep of cleanups.

File: traits/observers/expression.py (hunk 6 above)
Comment (mdickinson) [reply traits/observers/expression.py:122]: Not sure if you missed this one, but it's still outstanding.

Comment (kitchoi) [reply traits/observers/expression.py]: Thanks!

**Comment by mdickinson** [comment]
LGTM
